core/services/keeper/upkeep_executer.go

```diff
@@ -0,0 +1,189 @@
+package keeper
+
+import (
+	"context"
+	"errors"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/store"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+	"go.uber.org/atomic"
+	"gorm.io/gorm"
+)
+
+const (
+	checkUpkeep        = "checkUpkeep"
+	performUpkeep      = "performUpkeep"
+	executionQueueSize = 10
+)
+
+func NewUpkeepExecuter(
+	db *gorm.DB,
+	ethClient eth.Client,
+) UpkeepExecuter {
+	return UpkeepExecuter{
+		blockHeight:    atomic.NewInt64(0),
+		ethClient:      ethClient,
+		keeperORM:      NewORM(db),
+		isRunning:      atomic.NewBool(false),
+		executionQueue: make(chan struct{}, executionQueueSize),
+		chDone:         make(chan struct{}),
+		chSignalRun:    make(chan struct{}, 1),
+	}
+}
+
+// UpkeepExecuter fulfills HeadTrackable interface
+var _ store.HeadTrackable = UpkeepExecuter{}
+
+type UpkeepExecuter struct {
+	blockHeight *atomic.Int64
+	ethClient   eth.Client
+	keeperORM   KeeperORM
+	isRunning   *atomic.Bool
+
+	executionQueue chan struct{}
+	chDone         chan struct{}
+	chSignalRun    chan struct{}
+}
+
+func (executer UpkeepExecuter) Start() error {
+	if executer.isRunning.Load() {
+		return errors.New("already started")
+	}
+	executer.isRunning.Store(true)
+	go executer.run()
+	return nil
+}
+
+func (executer UpkeepExecuter) Close() error {
+	close(executer.chDone)
+	return nil
+}
+
+func (executer UpkeepExecuter) Connect(head *models.Head) error {
+	return nil
+}
+
+func (executer UpkeepExecuter) Disconnect() {}
+
+func (executer UpkeepExecuter) OnNewLongestChain(ctx context.Context, head models.Head) {
+	executer.blockHeight.Store(head.Number)
+	// avoid blocking if signal already in buffer
+	select {
+	case executer.chSignalRun <- struct{}{}:
+	default:
+	}
+}
+
+func (executer UpkeepExecuter) run() {
+	for {
+		select {
+		case <-executer.chDone:
+			return
+		case <-executer.chSignalRun:
+			executer.processActiveUpkeeps()
+		}
+	}
+}
+
+func (executer UpkeepExecuter) processActiveUpkeeps() {
+	// Keepers could miss their turn in the turn taking algo if they are too overloaded
+	// with work because processActiveUpkeeps() blocks
+	logger.Debug("received new block, running checkUpkeep for keeper registrations")
```

Comment:
Let's add more structure logging here, it's always helpful e.g. block height etc